temporalio/worker/workflow.py

```diff
@@ -0,0 +1,251 @@
+
+logger = logging.getLogger(__name__)
+
+LOG_PROTOS = False
+DEADLOCK_TIMEOUT_SECONDS = 2
+
+
+class _WorkflowWorker:
+    def __init__(
+        self,
+        *,
+        bridge_worker: Callable[[], temporalio.bridge.worker.Worker],
+        namespace: str,
+        task_queue: str,
+        workflows: Iterable[Type],
+        workflow_task_executor: Optional[concurrent.futures.ThreadPoolExecutor],
+        workflow_runner: WorkflowRunner,
+        data_converter: temporalio.converter.DataConverter,
+        interceptors: Iterable[Interceptor],
+        type_hint_eval_str: bool,
+        max_concurrent_workflow_tasks: int,
+    ) -> None:
+        self._bridge_worker = bridge_worker
+        self._namespace = namespace
+        self._task_queue = task_queue
+        self._workflow_task_executor = (
+            workflow_task_executor
+            or concurrent.futures.ThreadPoolExecutor(
+                max_workers=max_concurrent_workflow_tasks,
+                thread_name_prefix="temporal_workflow_",
+            )
+        )
+        self._workflow_task_executor_user_provided = workflow_task_executor is not None
+        self._workflow_runner = workflow_runner
+        self._data_converter = data_converter
+        self._interceptor_classes: List[Type[WorkflowInboundInterceptor]] = []
+        for i in interceptors:
+            interceptor_class = i.workflow_interceptor_class()
+            if interceptor_class:
+                self._interceptor_classes.append(interceptor_class)
+        self._type_hint_eval_str = type_hint_eval_str
+        self._running_workflows: Dict[str, WorkflowInstance] = {}
+
+        # Validate and build workflow dict
+        self._workflows: Dict[str, temporalio.workflow._Definition] = {}
+        for workflow in workflows:
+            defn = temporalio.workflow._Definition.must_from_class(workflow)
+            # Confirm name unique
+            if defn.name in self._workflows:
+                raise ValueError(f"More than one workflow named {defn.name}")
+            self._workflows[defn.name] = defn
+
+    async def run(self) -> None:
+        # Continually poll for workflow work
+        task_tag = object()
+        try:
+            while True:
+                act = await self._bridge_worker().poll_workflow_activation()
+                # Schedule this as a task, but we don't need to track it or
+                # await it. Rather we'll give it an attribute and wait for it
+                # when done.
+                task = asyncio.create_task(self._handle_activation(act))
+                setattr(task, "__temporal_task_tag", task_tag)
+        except temporalio.bridge.worker.PollShutdownError:
+            return
+        except Exception:
+            # Should never happen
+            logger.exception(f"Workflow runner failed")
+        finally:
+            # Collect all tasks and wait for them to complete
+            our_tasks = [
+                t
+                for t in asyncio.all_tasks()
```

Comment:
Why not store the task refs in the worker instance?

Reply:
It wouldn't be in the worker instance per se, it'd be here on the worker. But tasks come and go when completed. It is easier to ask the event loop for all running tasks than it is to maintain our own list of running tasks that I'd have to add here and then remove at the end of `_handle_activation` complete.

Reply:
It is possible that's a concern since this is on an event loop that is shared by potentially thousands of workers. But yeah, since this is on shutdown only it shouldn't be _too_ bad. And technically it's only all outstanding tasks which shouldn't be _too_ large at any one time.

Reply:
👍